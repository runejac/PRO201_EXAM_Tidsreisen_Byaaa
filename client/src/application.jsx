import { BrowserRouter, Route, Routes } from "react-router-dom";
import { LoginPage } from "./pages/LoginPage.jsx";
import { MapPage } from "./pages/MapPage.jsx";
import { Quiz } from "./pages/capsules/quizCapsule/Quiz.jsx";
import { History } from "./pages/capsules/HistoryCapsule/History.jsx";
import { MyFindings } from "./pages/MyFindings.jsx";
import { UserContext } from "./contexts/userContext.jsx";
import { Profile } from "./pages/Profile.jsx";
import { NotFound } from "./components/NotFound.jsx";
import { BackButton } from "./components/BackButton";
import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";
import { useLoading } from "./helpers/useLoading.jsx";
import WithoutNavbar from "./components/WithoutNavbar";
import WithNavbar from "./components/WithNavbar";
import Audio from "./pages/capsules/AudioCapsule/Audio.jsx";
import Camera from "./components/Camera";
import WithBackButton from "./components/WithBackButton.jsx";

export const User = createContext("");
export const MapContext = React.createContext(null);

export function Application() {
  const [user, setUser] = useState([]);
  const [map, setMap] = useState(null);

  const { getUser } = useContext(UserContext);
  const { data: username, reload, loading, error } = useLoading(getUser);

  const providerValue = useMemo(() => ({ user, setUser }), [user, setUser]);
  const mapProvider = useMemo(() => ({ map, setMap }), [map, setMap]);

  useEffect(() => {
    username ? setUser(...username, username) : reload();
  }, [username]);

  return (
    <User.Provider value={providerValue}>
      <MapContext.Provider value={mapProvider}>
        <BrowserRouter>
          <main>
            <Routes>
              <Route element={<WithoutNavbar />}>
                <Route path="/" element={<LoginPage />} />
              </Route>
              <Route element={<WithNavbar />}>
                <Route path="/map" element={<MapPage />} />
                <Route path="/camera" element={<Camera />} />
                <Route element={<WithBackButton />}>
                  <Route path="/quiz/:id" element={<Quiz />} />
                  <Route path="/history/:id" element={<History />} />
                  <Route
                    path="/myfindings"
                    element={<MyFindings loading={loading} error={error} />}
                  />
                  <Route path="/audio/:id" element={<Audio />} />
                  <Route path="/profile" element={<Profile />} />
                  <Route path={"*"} element={<NotFound />} />
                </Route>
              </Route>
            </Routes>
          </main>
        </BrowserRouter>
      </MapContext.Provider>
    </User.Provider>
  );
}
