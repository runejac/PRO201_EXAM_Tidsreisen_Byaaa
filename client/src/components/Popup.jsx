import React, { useContext, useState } from "react";
import Modal from "react-bootstrap/Modal";
import { CapsuleButtonGreen } from "./CapsuleButton.jsx";
import { BsArrowRight } from "react-icons/bs";
import "../css/popup.css";
import olaf from "../assets/images/olaf-infront-intro.png";
import { postJSON } from "../helpers/http.jsx";
import { User } from "../application.jsx";

export default function Popup() {
  const { user, setUser } = useContext(User);
  const { name, intro } = user;
  const [show, setShow] = useState(true);

  const [message, setMessage] = useState(() => {
    return (
      <div style={{ marginTop: "1rem" }}>
        Trykk på <BsArrowRight /> for å lære litt om appen
      </div>
    );
  });

  const messages = [
    {
      id: 1,
      content: (
        <p className={"intro-content"}>
          <span style={{ fontWeight: "bold" }}>
            Heisann {name}, jeg heter Olaf! <br />
          </span>
          Jeg jobber som sagdreng her på Byåa. Bli med meg på denne tidsreisen!
        </p>
      ),
    },
    {
      id: 2,
      content: (
        <p className={"intro-content"}>
          Reisen begynner på kapsel nummer 1 på kartet. Bruk bryteren som er
          øverst til høyre for å velge om du skal bruke appen på skolen, eller
          på stien. Klikk på kartet, eller på QR-scan ikonet som du finner på
          midten i menyen.
        </p>
      ),
    },
    {
      id: 3,
      content: (
        <p className={"intro-content"}>
          Jeg har gjemt forskjellige tidskapsler. Beveg deg videre på
          kulturstien, og reis tilbake i tid med meg.
          <br />
          I tillegg får du poeng for hver tidskapsel du fullfører. Håper du
          finner dem alle.
          <br />
          Husk å ha det gøy og lykke til {name}!
        </p>
      ),
    },
  ];

  const [next, setNext] = useState(1);

  async function changeTitle(id) {
    if (id > 3) await handleClose();

    messages.map((modalMessage) => {
      if (modalMessage.id === id) {
        setMessage(modalMessage.content);
      }
    });
  }

  let previousState = { ...user };

  const handleClose = async () => {
    await postJSON("/api/update-state", {
      user: name,
      intro: false,
    });
    setShow(false);

    previousState.intro = false;
    setUser({ ...previousState });
  };

  const handleNext = () => setNext(next + 1);
  const handleShow = () => setShow(true);

  if (intro) window.addEventListener("load", handleShow);

  return (
    <section className={"rounded"}>
      {name && (
        <Modal
          dialogClassName={"border-radius-2"}
          className={"d-flex justify-content-center align-items-center"}
          show={show}
          onHide={handleClose}
        >
          <Modal.Body id={"modal-body"}>
            <div id={"modal-title"}>
              <h1 id={"title"}>Velkommen til Tidsreisen!</h1>
            </div>

            <div id={"image-and-message-container"}>
              <img
                width={200}
                id={"olaf-image"}
                src={olaf}
                alt="bilde av intro-olaf"
              />
              <span>{message}</span>
            </div>

            <div id={"bottom-container"}>
              <div className={"d-flex flex-column"}>
                {next > 3 && (
                  <CapsuleButtonGreen
                    className={"justify-content-center align-content-center"}
                    buttonText={"Start reisen!"}
                    onClick={handleClose}
                  />
                )}
              </div>
              {3 >= next ? (
                <BsArrowRight
                  style={{
                    fontSize: "2.5rem",
                    color: "var(--backgroundColorGreeny)",
                  }}
                  onClick={() => {
                    handleNext();
                    changeTitle(next);
                  }}
                />
              ) : (
                ""
              )}
            </div>
          </Modal.Body>
        </Modal>
      )}
    </section>
  );
}
