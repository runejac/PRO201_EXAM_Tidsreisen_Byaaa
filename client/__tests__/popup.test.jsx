import React from "react";
import { MemoryRouter } from "react-router-dom";
import Popup from "../src/components/Popup.jsx";
import renderer from "react-test-renderer";
import { User } from "../src/application.jsx";

describe("Popup", () => {
  it("shows snapshot", async () => {
    const user = { name: "Hei", intro: false };
    const component = renderer.create(
      <MemoryRouter>
        <User.Provider value={{ user }}>
          <Popup />
        </User.Provider>
      </MemoryRouter>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
