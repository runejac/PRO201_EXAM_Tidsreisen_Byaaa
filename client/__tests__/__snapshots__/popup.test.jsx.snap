// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup shows snapshot 1`] = `
<section
  className="rounded"
/>
`;
